src/types/any.rs

```diff
@@ -99,6 +99,9 @@ impl PyAny {
     /// Determines whether this object has the given attribute.
     ///
     /// This is equivalent to the Python expression `hasattr(self, attr_name)`.
+    ///
+    /// To avoid repeated temporary allocations of Python strings, the [`intern!`] macro can be used
+    /// to intern `attr_name`.
     pub fn hasattr<N>(&self, attr_name: N) -> PyResult<bool>
     where
         N: ToPyObject,
```

Comment:
As things are, this would leave us with two different kinds of bounds for names, either `N: ToPyObject` or `N: IntoPy<Py<PyString>>`. Would it make sense to unify them?

Reply:
I looked into it and I think attribute names don't need to be strings - anything will do as long as it inherits from `str`.

I would be happy to limit it to `IntoPy<Py<PyString>>`, but - are people relying on this behaviour?

For example:
```rust
use pyo3::prelude::*;

#[pyfunction]
fn getattr(item: &PyAny, arg: &PyAny) -> PyResult<()> {
    item.getattr(arg)?;
    Ok(())
}

/// This module is implemented in Rust.
#[pymodule]
fn my_module(py: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(getattr, m)?)?;
    Ok(())
}
```

```python
import my_module

class A:
    def __getattr__(self, attr):
        print(attr)
        print(attr.__class__)
        return "foo"

a = A()

class C(str):
    pass

my_module.getattr(a, "field")
my_module.getattr(a, C())
```

Reply:
I think `IntoPy<Py<PyString>>` makes sense for attribute names. Edge cases like this don't seem likely to me, and I think with a bit of type wrangling (e.g. change `arg: &PyAny` to `arg: &PyString`) things should still work.

Reply:
Done